
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <title>mutationseq | Shah Lab</title>
    
    <meta name="author" content="Sohrab P. Shah">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>

    <!-- Bootstrap styles -->
    <link href="/assets/themes/lab/bootstrap/css/bootstrap.min.css" rel="stylesheet"/>
    <!-- Optional theme -->
    <link href="/assets/themes/lab/bootstrap/css/bootstrap-theme.min.css" rel="stylesheet"/>
    <!-- Sticky Footer -->
    <link href="/assets/themes/lab/bootstrap/css/bs-sticky-footer.css" rel="stylesheet"/>
    
    <!-- Custom styles -->
    <link href="/assets/themes/lab/css/style.css?body=1" rel="stylesheet" type="text/css" media="all"/>

    <!-- icon -->
    <link rel="icon" type="image/png" href="/assets/themes/lab/images/logo/favicon.png"/>

    <!-- Custom fonts and icons via Font Awesome, http://fortawesome.github.io/Font-Awesome/ -->
    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css"/>

    <!-- Fonts via Google -->
    <link href='http://fonts.googleapis.com/css?family=Lato:300italic,700italic,300,700' rel='stylesheet' type='text/css'/>

    <!-- Math via MathJax -->
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
      <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->

    <!-- Fav and touch icons -->
    <!-- Update these with your own images
      <link rel="shortcut icon" href="images/favicon.ico">
      <link rel="apple-touch-icon" href="images/apple-touch-icon.png">
      <link rel="apple-touch-icon" sizes="72x72" href="images/apple-touch-icon-72x72.png">
      <link rel="apple-touch-icon" sizes="114x114" href="images/apple-touch-icon-114x114.png">
    -->

    <!-- atom & rss feed
    <link href="/atom.xml" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="/rss.xml" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">
    -->

  </head>

  <body>

  <!-- Static top navbar -->

  <nav class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class='container'>
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="navbar-header">
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#jb-navbar-collapse">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a href="/">
        <img class="pull-left logo" src="/assets/themes/lab/images/logo/lab-logo.svg" type="image/svg+xml">
        </a>
      </div>

      <!-- Collect the nav links, forms, and other content for toggling -->
      <div class="collapse navbar-collapse" id="jb-navbar-collapse">
        <ul class="nav navbar-nav navbar-left hidden-sm">
          <!-- Don't put site title on home page -- there's already a big title there. -->
          
          
          
          <a class="navbar-brand" href="/">Shah Lab</a>
          
        </ul>
        <ul class="nav navbar-nav navbar-right pull-right">
          
          
          


  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/research">Research</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/blog/">Blog</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/papers/">Papers</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/projects/">Software</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/team/">Team</a></li>
      	
      
    
  



          <!--<li class="dropdown nav navbar-nav">-->
            <!--<a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">-->
              <!--More <span class="caret"></span>-->
            <!--</a>-->

            <!--<ul class="dropdown-menu nav navbar-nav" role="menu">-->
              <!---->
              <!---->
              <!--


  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  


-->
            <!--</ul>-->
          <!--</li>-->
        </ul>
      </div><!-- /.navbar-collapse -->
    </nav>
   </div>

   <!-- Content -->
   <div class="container">
      
<div class="page-header">
  <div class="title media-heading">mutationseq</div>
  
</div>

<div class="row post-full">
  <div class="col-xs-12">
    <div class="date">
      <span></span>
    </div>
    <div class="content">
      <p><a href="https://github.com/shahcompbio/mutationseq">Fork me on GitHub!</a></p>

<h3 id="about">About</h3>
<p>mutationSeq is software for somatic SNV detection using next generation sequencing (NGS) data from matched tumour/normal samples. It uses
a feature-based classifier trained on validated somatic mutation samples while benefiting from other available information such as base quality,
mapping quality, strand bias and tail distance.</p>

<h3 id="note">Note</h3>
<p><code class="highlighter-rouge">model_single_v4.0.1.npz</code> and <code class="highlighter-rouge">model_single_v4.0.2_anaconda_sk_0.13.1.npz</code> are missing from this repository, and can be found at <code class="highlighter-rouge">ftp://ftp.bcgsc.ca/public/shahlab/MutationSeq/</code>. Once youâ€™ve acquired those files, place <code class="highlighter-rouge">model_single_v4.0.1.npz</code> in the root of this repo, and place <code class="highlighter-rouge">model_single_v4.0.2_anaconda_sk_0.13.1.npz</code> in the <code class="highlighter-rouge">models_anaconda</code> directory.</p>

<h3 id="installing-mutationseq">Installing MutationSeq</h3>

<p>MutationSeq can be downloaded from http://compbio.bccrc.ca/software/mutationseq or https://bitbucket.org/shahlabbcca/mutationseq. Once you have downloaded it, extract it:</p>

<pre><code class="language-{bash}">tar -xzvf museq_4.3.8.tar.gz
</code></pre>

<p>This will extract the content into a folder <code class="highlighter-rouge">mutationseq</code>. We will move the files into <code class="highlighter-rouge">$HOME/usr/museq/4.3.8</code> for organization purposes:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>mkdir -p $HOME/usr/museq/4.3.8
mv mutationseq/* $HOME/usr/museq/4.3.8
</code></pre>
</div>

<p>Now we need to install MutationSeq. MutationSeq requires python (v2.7) and several key package dependencies.</p>

<ul>
  <li>numpy</li>
  <li>scipy</li>
  <li>matplotlib</li>
  <li>scikit-learn</li>
  <li>intervaltree</li>
</ul>

<p>The best way to install all of this is to use either <a href="http://conda.pydata.org/miniconda.html">Miniconda</a> or anaconda. We will use miniconda here. First download miniconda (for python 2.7) and then run:</p>

<pre><code class="language-{bash}">sh Miniconda2-latest-Linux-x86_64.sh
</code></pre>

<p>Then follow the instructions. When you have finished following the instructions, you should have python installed:</p>

<pre><code class="language-{bash}">which python
~/miniconda2/bin/python
</code></pre>

<p>Now we can install the dependencies needed (NOTE: Versions and specified to ensure compatibility)</p>

<pre><code class="language-{bash}">conda install -c bioconda numpy=1.7.1 scipy=0.12.0 scikit-learn=0.13.1 matplotlib=1.2.1 intervaltree
</code></pre>

<p>One last thing that is needed before we can install MutationSeq is the Boost C libraries. We only need to download them from http://www.boost.org/. Once you have downloaded (tested on 1.51) just extract them to a location. For example, you could put it into <code class="highlighter-rouge">$HOME/usr/boost/1.51</code></p>

<p>Once this has been installed, we can now proceed to compiling a dependency <code class="highlighter-rouge">pybam.so</code>.</p>

<pre><code class="language-{bash}">make PYTHON=python BOOSTPATH=$HOME/usr/boost/1.51 -B
</code></pre>

<p>Now when you run:</p>

<pre><code class="language-{bash}">python $HOME/usr/museq/4.3.8/museq/classify.py --version
4.3.8
</code></pre>

<p>This indicates that you have successfully installed MutationSeq.</p>

<blockquote>
  <p>An important thing to note is that MutationSeq comes bundled with a trained classifer using the <a href="http://scikit-learn.org/">scikit-learn</a> library and packaged as a pickle. The pickle is only compatible with the specific scikit-learn version that it was built with. As we are using miniconda here with scikit-learn version 0.13.1, we will have to use the corresponding models associated with that version. See <a href="#using-mutationseq">Calling Variants - Using MutationSeq</a> for more details.</p>
</blockquote>

<h3 id="using-mutationseq">Using MutationSeq</h3>

<p>To call variants using MutationSeq, we use the following command:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>mkdir -p museq/results; \
python $(HOME)/usr/museq/4.3.8/museq/classify.py \
  normal:bam/HCC1395_exome_normal.sort.markdup.17.7MB-8MB.bam \
  tumour:bam/HCC1395_exome_tumour.sort.markdup.17.7MB-8MB.bam \
  reference:refs/GRCh37-lite.fa \
  model:$(HOME)/museq/4.3.8/museq/models_anaconda/model_v4.1.2_anaconda_sk_0.13.1.npz \
  -c $(HOME)/usr/museq/4.3.8/museq/metadata.config \
  -o museq/results/HCC1395_exome_tumour_normal_17.vcf
</code></pre>
</div>

<p>Notice how for the <code class="highlighter-rouge">model</code> parameter, we used the <code class="highlighter-rouge">$(HOME)/museq/4.3.8/museq/models_anaconda/model_v4.1.2_anaconda_sk_0.13.1.npz</code>. As mentioned in the <a href="#installing-mutationseq">Installing MutationSeq</a> section, this model specifically works with conda and the scikit-learn library (v0.13.1) which is how we installed MutationSeq in the workshop. If you are not using conda, then you will have to use the model version which is compatible with the scikit-learn library you have installed for your python.</p>


    </div>

  <!-- 
    <ul class="tag_box inline">
      <li><i class="glyphicon glyphicon-open"></i></li>
      
      


  
    
  


    </ul>
    

  
    <ul class="tag_box inline">
      <li><i class="glyphicon glyphicon-tags"></i></li>
      
      


  
    
  



    </ul>
   -->  
  
    <hr>
    
  </div>
</div>


   </div>


  <!-- Static bottom navbar -->
  <nav class="navbar navbar-default navbar-fixed-bottom">
    <div class="container">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#jb-bottom-collapse"> <!-- NOTE! data-target was changed to jb-bottom-collapse -->
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
      </div>
      <!-- Collect the nav links, forms, and other content for toggling -->
      <div class="collapse navbar-collapse" id="jb-bottom-collapse">
        <a href=http://molonc.bccrc.ca>
          <img class="logo" alt="BCCA Molecular Oncology" src=/assets/images/bcca_logo.png>
        </a>
        <a href=http://www.ubc.ca/>
          <img class="logo" alt="UBC" src=/assets/images/4_2016_UBCLogo_BlueRGB72.png>
        </a>
        <ul class="nav navbar-nav navbar-right navbar-small">
          <li><a href='/about.html'>About</a></li>
          <li><a href='/join.html'>Join</a></li>
          <li><a href='/contact.html'>Contact</a></li>
        </ul>
      </div><!-- /.navbar-collapse -->
    </div>
  </nav>

    




  <script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-103919622-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>




    <!-- Latest compiled and minified JavaScript, requires jQuery 1.x (2.x not supported in IE8) -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
    <script src="/assets/themes/lab/bootstrap/js/bootstrap.min.js"></script>
    <script type='text/javascript' src='https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js'></script>
  </body>
</html>

