
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <title>TitanCNA | Shah Lab</title>
    
    <meta name="author" content="Sohrab P. Shah">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>

    <!-- Bootstrap styles -->
    <link href="/assets/themes/lab/bootstrap/css/bootstrap.min.css" rel="stylesheet"/>
    <!-- Optional theme -->
    <link href="/assets/themes/lab/bootstrap/css/bootstrap-theme.min.css" rel="stylesheet"/>
    <!-- Sticky Footer -->
    <link href="/assets/themes/lab/bootstrap/css/bs-sticky-footer.css" rel="stylesheet"/>
    
    <!-- Custom styles -->
    <link href="/assets/themes/lab/css/style.css?body=1" rel="stylesheet" type="text/css" media="all"/>

    <!-- icon -->
    <link rel="icon" type="image/png" href="/assets/themes/lab/images/logo/favicon.png"/>

    <!-- Custom fonts and icons via Font Awesome, http://fortawesome.github.io/Font-Awesome/ -->
    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css"/>

    <!-- Fonts via Google -->
    <link href='http://fonts.googleapis.com/css?family=Lato:300italic,700italic,300,700' rel='stylesheet' type='text/css'/>

    <!-- Math via MathJax -->
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
      <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->

    <!-- Fav and touch icons -->
    <!-- Update these with your own images
      <link rel="shortcut icon" href="images/favicon.ico">
      <link rel="apple-touch-icon" href="images/apple-touch-icon.png">
      <link rel="apple-touch-icon" sizes="72x72" href="images/apple-touch-icon-72x72.png">
      <link rel="apple-touch-icon" sizes="114x114" href="images/apple-touch-icon-114x114.png">
    -->

    <!-- atom & rss feed
    <link href="/atom.xml" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="/rss.xml" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">
    -->

  </head>

  <body>

  <!-- Static top navbar -->

  <nav class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class='container'>
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="navbar-header">
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#jb-navbar-collapse">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a href="/">
        <img class="pull-left logo" src="/assets/themes/lab/images/logo/lab-logo.svg" type="image/svg+xml">
        </a>
      </div>

      <!-- Collect the nav links, forms, and other content for toggling -->
      <div class="collapse navbar-collapse" id="jb-navbar-collapse">
        <ul class="nav navbar-nav navbar-left hidden-sm">
          <!-- Don't put site title on home page -- there's already a big title there. -->
          
          
          
          <a class="navbar-brand" href="/">Shah Lab</a>
          
        </ul>
        <ul class="nav navbar-nav navbar-right pull-right">
          
          
          


  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/research">Research</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/blog/">Blog</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/papers/">Papers</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/projects/">Software</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/team/">Team</a></li>
      	
      
    
  



          <!--<li class="dropdown nav navbar-nav">-->
            <!--<a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">-->
              <!--More <span class="caret"></span>-->
            <!--</a>-->

            <!--<ul class="dropdown-menu nav navbar-nav" role="menu">-->
              <!---->
              <!---->
              <!--


  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  


-->
            <!--</ul>-->
          <!--</li>-->
        </ul>
      </div><!-- /.navbar-collapse -->
    </nav>
   </div>

   <!-- Content -->
   <div class="container">
      
<div class="page-header">
  <div class="title media-heading">TitanCNA</div>
  
</div>

<div class="row post-full">
  <div class="col-xs-12">
    <div class="date">
      <span></span>
    </div>
    <div class="content">
      <p><a href="https://github.com/gavinha/TitanCNA">Fork me on GitHub!</a></p>

<p><a href="https://travis-ci.org/gavinha/TitanCNA"><img src="https://travis-ci.org/gavinha/TitanCNA.svg?branch=master" alt="Build Status" /></a></p>

<h1 id="titancna"><em>TitanCNA</em></h1>

<p>TitanCNA a R/Bioconductor package for analyzing subclonal copy number alterations (CNA) and loss of heterozygosity (LOH) in whole genome and exome sequencing of tumours.</p>

<p>Ha, G., et al. (2014). <a href="http://genome.cshlp.org/content/24/11/1881">TITAN: Inference of copy number architectures in clonal cell populations from tumour whole genome sequence data. Genome Research, 24: 1881-1893.</a> (PMID: 25060187)</p>

<h2 id="contact">Contact</h2>
<p>Gavin Ha<br />
Fred Hutchinson Cancer Research Center<br />
contact: <a href="mailto:gavinha@gmail.com">gavinha@gmail.com</a> or <a href="mailto:gha@fredhutch.org">gha@fredhutch.org</a><br />
Date: August 9, 2018<br />
Website: <a href="https://gavinhalab.org/">GavinHaLab.org</a></p>

<h2 id="table-of-contents">Table of Contents</h2>
<ul>
  <li><a href="#links">Links</a></li>
  <li><a href="#news">News</a></li>
  <li><a href="#installation">Installation</a></li>
  <li><a href="#usage">Usage</a></li>
  <li><a href="#vignette-in-titancna-r-package">Vignette in TitanCNA R package</a></li>
  <li><a href="#acknowledgements">Acknowledgements</a></li>
  <li><a href="#software-license">License</a></li>
</ul>

<h2 id="links">Links</h2>
<p>Snakemake Workflow: https://github.com/gavinha/TitanCNA/tree/master/scripts/snakemake<br />
<strong>10X Snakemake Workflow:</strong> https://github.com/gavinha/TitanCNA_10X_snakemake<br />
Google Groups: https://groups.google.com/forum/#!forum/titancna<br />
TitanCNA website: http://compbio.bccrc.ca/software/titan/<br />
KRONOS TITAN Workflow: https://github.com/MO-BCCRC/titan_workflow<br />
Publication in Genome Research: http://genome.cshlp.org/content/24/11/1881</p>

<h2 id="news">News</h2>
<p>(See <a href="/projects/TitanCNA/NEWS">NEWS</a> for previous version notes)</p>

<h3 id="august-9-2018">August 9, 2018</h3>
<p>Improved parameter inference by handling errors and allowing EM to continue until convergence. This fixes runs that previously would fail because samples had very low tumor content.</p>

<h3 id="july-26-2018">July 26, 2018</h3>
<p>Snakemake workflow for 10X Genomics whole genome sequencing data is now included in another Git repo.
https://github.com/gavinha/TitanCNA_10X_snakemake</p>

<h3 id="titancna-version-1171-changes">TitanCNA version 1.17.1 changes</h3>
<p>1)  New functions: 
	- <code class="highlighter-rouge">correctIntegerCN()</code>: recomputes high-level copy number that is capped by the maximum CN state. 
	Performs two tasks - (1) correct log ratio based on purity and ploidy, and then convert to decimal CN value; (2) Correct bins and segments in which the original predicted integer copy number was assigned the maximum CN state; bins and segments for all of chromosome X are also corrected, if provided in the input.</p>

<p>2) Modified functions:
	- <code class="highlighter-rouge">plotSegmentMedians()</code> and <code class="highlighter-rouge">plotCNlogRByChr()</code>: includes argument to show color-coding for corrected copy number; defaults to TRUE for this argument.</p>

<p>3) Removed functions/manual/dependencies:
	- <code class="highlighter-rouge">extractAlleleReadCounts()</code>	
	- <code class="highlighter-rouge">Rsamtools</code> dependency</p>

<h3 id="titancna-version-1150-changes">TitanCNA version 1.15.0 changes</h3>
<p>1) 10X Genomics analysis</p>
<ul>
  <li>Please see <a href="scripts/">scripts</a> for instructions on running the 10X Genomics analysis.</li>
</ul>

<p>2) New script to help <strong>select optimal solutions</strong>.<br />
	Please see <a href="https://github.com/gavinha/TitanCNA/tree/master/scripts/R_scripts">scripts/R_scripts</a></p>

<p>3) Added snakemake pipeline for entire TITAN workflow<br />
	Please see <a href="scripts/snakemake">scripts/snakemake</a>.</p>

<p>4) New function:
	- <code class="highlighter-rouge">plotSegmentMedians()</code>
	- <code class="highlighter-rouge">loadHaplotypeAlleleCounts()</code>: loads input allele counts with phasing information
	- <code class="highlighter-rouge">plotHaplotypeFraction()</code>: results from 10X Genomics WGS data with phasing of haplotype blocks</p>

<p>5) Modified features (no changes for user-accessible functions):
	- updateParameters: coordinate descent estimate of ploidy update uses previously estimated normal parameter from the same corodinate descent iteration ; leads to faster convergence</p>

<h2 id="installation">Installation</h2>
<h3 id="install-titancna-r-package-from-github">Install TitanCNA R package from github</h3>

<p>From within R-3.3.2 or higher,</p>
<div class="highlighter-rouge"><pre class="highlight"><code>install.packages("devtools")
library(devtools)
install_github("gavinha/TitanCNA")
</code></pre>
</div>

<h3 id="install-titancna-from-bioconductor">Install TitanCNA from Bioconductor</h3>
<p>From within R-3.3.2 or higher,</p>
<div class="highlighter-rouge"><pre class="highlight"><code>source("https://bioconductor.org/biocLite.R")
biocLite("TitanCNA")
</code></pre>
</div>

<h3 id="install-other-dependencies">Install other dependencies</h3>
<ol>
  <li>
    <p>Install the HMMcopy suite
Please follow instructions on the HMMcopy website <a href="http://compbio.bccrc.ca/software/hmmcopy/">http://compbio.bccrc.ca/software/hmmcopy/</a>.</p>
  </li>
  <li>
    <p><a href="https://github.com/MO-BCCRC/titan_workflow">KRONOS</a> TITAN Workflow
The easiest way to generate these files is by using the downloadable pipeline from https://github.com/MO-BCCRC/titan_workflow.</p>
  </li>
</ol>

<h2 id="usage">Usage</h2>
<p>R scripts are provided to run the R component of the TITAN analysis using the TitanCNA R/Bioconductor package.<br />
Please go to the <a href="scripts/">scripts</a> directory and look at the README there for more details.</p>

<h3 id="snakemake-workflow">Snakemake workflow</h3>
<p>A <a href="scripts/snakemake">snakemake</a> is also provided in this repo.<br />
This workflow will run the TITAN a set of tumour-normal pairs, starting from the BAM files and generating TitanCNA outputs. It will also perform model selection at the end of the workflow to choose the optimal ploidy and clonal cluster solutions.</p>

<h2 id="vignette-in-titancna-r-package">Vignette in TitanCNA R package</h2>
<p>The PDF of the vignette can be accessed from R</p>
<div class="highlighter-rouge"><pre class="highlight"><code>library(TitanCNA)
browseVignettes(package = "TitanCNA")
</code></pre>
</div>
<p>The path of the file can also be located using</p>
<div class="highlighter-rouge"><pre class="highlight"><code>pathToInstall &lt;- system.file(package = "TitanCNA")
pathToPdf &lt;- paste0(pathToInstall, "/int/doc/TitanCNA.pdf)
</code></pre>
</div>
<p>The example provided will reproduce Figure 1 in the manuscript. However, it will be slightly different because the example is only based on the analysis of chr2, not genome-wide.</p>

<h2 id="acknowledgements">Acknowledgements</h2>
<p>TitanCNA was developed by Gavin Ha while in the laboratories of Sohrab Shah (sshah@bccrc.ca) and Sam Aparicio (saparicio@bccrc.ca) at the Dept of Molecular Oncology, BC Cancer Agency, Vancouver, Canada.<br />
Yikan Wang and Daniel Lai have contributed code and discussions to this project.<br />
The KRONOS TITAN workflow was developed by Diljot Grewal (<a href="mailto:dgrewal@bccrc.ca">dgrewal@bccrc.ca</a>) and Jafar Taghiyar (<a href="mailto:jtaghiyar@bccrc.ca">jtaghiyar@bccrc.ca</a>).<br />
HMMcopy was co-developed by Daniel Lai and Gavin Ha.</p>

<p>TitanCNA was inspired by existing methods including <a href="https://sites.google.com/site/oncosnp/">OncoSNP</a>  and <a href="https://bitbucket.org/aroth85/pyclone/wiki/Home">PyClone</a></p>

<h2 id="software-license">Software License</h2>
<p>License: <a href="/projects/TitanCNA/LICENSE">GPLv3</a></p>

<p>TitanCNA R code is open source and R/Bioconductor package is under <a href="/projects/TitanCNA/LICENSE">GPLv3</a>.  This applies to the v1.9.0 and all subsequent versions within and obtained from Bioconductor.</p>

<p>Users who are using TitanCNA earlier than v1.9.0 not for the purpose of academic research should contact gavinha@broadinstitute.org, sshah@bccrc.ca, and prebstein@bccancer.bc.ca to inquire about previous licensing.</p>

    </div>

  <!-- 
    <ul class="tag_box inline">
      <li><i class="glyphicon glyphicon-open"></i></li>
      
      


  
    
  


    </ul>
    

  
    <ul class="tag_box inline">
      <li><i class="glyphicon glyphicon-tags"></i></li>
      
      


  
    
  



    </ul>
   -->  
  
    <hr>
    
  </div>
</div>


   </div>


  <!-- Static bottom navbar -->
  <nav class="navbar navbar-default navbar-fixed-bottom">
    <div class="container">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#jb-bottom-collapse"> <!-- NOTE! data-target was changed to jb-bottom-collapse -->
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
      </div>
      <!-- Collect the nav links, forms, and other content for toggling -->
      <div class="collapse navbar-collapse" id="jb-bottom-collapse">
        <a href=http://molonc.bccrc.ca>
          <img class="logo" alt="BCCA Molecular Oncology" src=/assets/images/bcca_logo.png>
        </a>
        <a href=http://www.ubc.ca/>
          <img class="logo" alt="UBC" src=/assets/images/4_2016_UBCLogo_BlueRGB72.png>
        </a>
        <ul class="nav navbar-nav navbar-right navbar-small">
          <li><a href='/about.html'>About</a></li>
          <li><a href='/join.html'>Join</a></li>
          <li><a href='/contact.html'>Contact</a></li>
        </ul>
      </div><!-- /.navbar-collapse -->
    </div>
  </nav>

    




  <script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-103919622-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>




    <!-- Latest compiled and minified JavaScript, requires jQuery 1.x (2.x not supported in IE8) -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
    <script src="/assets/themes/lab/bootstrap/js/bootstrap.min.js"></script>
    <script type='text/javascript' src='https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js'></script>
  </body>
</html>

