#' Tripple Negative breast cancer Xenograft, sample SA494, timepoint T
#'
#' A dataset containing the allele counts and genotypes from a Tripple negative breast cancer study (see
#' \cite{Eirew P, Steif A, Khattra J, Ha G, Yap D, Farahani H, et al. Dynamics of genomic clones in breast cancer patient xenografts at single-cell resolution. Nature. 2014;.})
#'
#' @format A list of 5 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'  }
#'
"tnbc.xenograft.SA494.T"


#' Tripple Negative breast cancer Xenograft, sample SA494, timepoint X4
#'
#' A dataset containing the allele counts and genotypes from a Tripple negative breast cancer study (see
#' \cite{Eirew P, Steif A, Khattra J, Ha G, Yap D, Farahani H, et al. Dynamics of genomic clones in breast cancer patient xenografts at single-cell resolution. Nature. 2014;.})
#'
#' @format A list of 5 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'  }
#'
"tnbc.xenograft.SA494.X4"

#' Tripple Negative breast cancer Xenograft, sample SA501, timepoint X1
#'
#' A dataset containing the allele counts and genotypes from a Tripple negative breast cancer study (see
#' \cite{Eirew P, Steif A, Khattra J, Ha G, Yap D, Farahani H, et al. Dynamics of genomic clones in breast cancer patient xenografts at single-cell resolution. Nature. 2014;.})
#'
#' @format A list of 5 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'  }
#'
"tnbc.xenograft.SA501.X1"

#' Tripple Negative breast cancer Xenograft, sample SA501, timepoint X2
#'
#' A dataset containing the allele counts and genotypes from a Tripple negative breast cancer study (see
#' \cite{Eirew P, Steif A, Khattra J, Ha G, Yap D, Farahani H, et al. Dynamics of genomic clones in breast cancer patient xenografts at single-cell resolution. Nature. 2014;.})
#'
#' @format A list of 5 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'  }
#'
"tnbc.xenograft.SA501.X2"

#' Tripple Negative breast cancer Xenograft, sample SA501, timepoint X4
#'
#' A dataset containing the allele counts and genotypes from a Tripple negative breast cancer study (see
#' \cite{Eirew P, Steif A, Khattra J, Ha G, Yap D, Farahani H, et al. Dynamics of genomic clones in breast cancer patient xenografts at single-cell resolution. Nature. 2014;.})
#'
#' @format A list of 5 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'  }
#'
"tnbc.xenograft.SA501.X4"




#' Simulated data from Generalized gd model, with seed 0
#'
#' A dataset containing the allele counts and genotypes from the Generalized gd model.
#' See (Salehi, Sohrab, Adi Steif, Roth Andrew, Samuel Aparicio, Alexandre Bouchard, and Sohrab P. Shah. “Joint Statistical Inference of Clonal Populations ￼from Single Cell and Bulk Tumour Sequencing Data.” (submitted).)
#'
#' @format A list of 8 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'   \item phi: For internal use of the package.
#'   \item trueMutMatrix: For internal use of the package.
#'   \item psis.single.cell: For internal use of the package.
#'   \item gdDirName: For internal use fo the package.
#'  }
#'
"gd.10.48.0.f0.gl0.u.tips.ADO.00.doublet.00.m.50.nus.0.da"


#' Simulated data from Generalized gd model, with seed 1
#'
#' A dataset containing the allele counts and genotypes from the Generalized gd model.
#' See (Salehi, Sohrab, Adi Steif, Roth Andrew, Samuel Aparicio, Alexandre Bouchard, and Sohrab P. Shah. “Joint Statistical Inference of Clonal Populations ￼from Single Cell and Bulk Tumour Sequencing Data.” (submitted).)
#'
#' @format A list of 8 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'   \item phi: For internal use of the package.
#'   \item trueMutMatrix: For internal use of the package.
#'   \item psis.single.cell: For internal use of the package.
#'   \item gdDirName: For internal use fo the package.
#'  }
#'
"gd.10.48.1.f0.gl0.u.tips.ADO.00.doublet.00.m.50.nus.0.da"


#' Simulated data from Generalized gd model, with seed 2
#'
#' A dataset containing the allele counts and genotypes from the Generalized gd model.
#' See (Salehi, Sohrab, Adi Steif, Roth Andrew, Samuel Aparicio, Alexandre Bouchard, and Sohrab P. Shah. “Joint Statistical Inference of Clonal Populations ￼from Single Cell and Bulk Tumour Sequencing Data.” (submitted).)
#'
#' @format A list of 8 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'   \item phi: For internal use of the package.
#'   \item trueMutMatrix: For internal use of the package.
#'   \item psis.single.cell: For internal use of the package.
#'   \item gdDirName: For internal use fo the package.
#'  }
#'
"gd.10.48.2.f0.gl0.u.tips.ADO.00.doublet.00.m.50.nus.0.da"


#' Simulated data from Generalized gd model, with seed 3
#'
#' A dataset containing the allele counts and genotypes from the Generalized gd model.
#' See (Salehi, Sohrab, Adi Steif, Roth Andrew, Samuel Aparicio, Alexandre Bouchard, and Sohrab P. Shah. “Joint Statistical Inference of Clonal Populations ￼from Single Cell and Bulk Tumour Sequencing Data.” (submitted).)
#'
#' @format A list of 8 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'   \item phi: For internal use of the package.
#'   \item trueMutMatrix: For internal use of the package.
#'   \item psis.single.cell: For internal use of the package.
#'   \item gdDirName: For internal use fo the package.
#'  }
#'
"gd.10.48.3.f0.gl0.u.tips.ADO.00.doublet.00.m.50.nus.0.da"


#' Simulated data from Generalized gd model, with seed 4
#'
#' A dataset containing the allele counts and genotypes from the Generalized gd model.
#' See (Salehi, Sohrab, Adi Steif, Roth Andrew, Samuel Aparicio, Alexandre Bouchard, and Sohrab P. Shah. “Joint Statistical Inference of Clonal Populations ￼from Single Cell and Bulk Tumour Sequencing Data.” (submitted).)
#'
#' @format A list of 8 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'   \item phi: For internal use of the package.
#'   \item trueMutMatrix: For internal use of the package.
#'   \item psis.single.cell: For internal use of the package.
#'   \item gdDirName: For internal use fo the package.
#'  }
#'
"gd.10.48.4.f0.gl0.u.tips.ADO.00.doublet.00.m.50.nus.0.da"


#' Simulated data from Generalized gd model, with seed 5
#'
#' A dataset containing the allele counts and genotypes from the Generalized gd model.
#' See (Salehi, Sohrab, Adi Steif, Roth Andrew, Samuel Aparicio, Alexandre Bouchard, and Sohrab P. Shah. “Joint Statistical Inference of Clonal Populations ￼from Single Cell and Bulk Tumour Sequencing Data.” (submitted).)
#'
#' @format A list of 8 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'   \item phi: For internal use of the package.
#'   \item trueMutMatrix: For internal use of the package.
#'   \item psis.single.cell: For internal use of the package.
#'   \item gdDirName: For internal use fo the package.
#'  }
#'
"gd.10.48.5.f0.gl0.u.tips.ADO.00.doublet.00.m.50.nus.0.da"


#' Simulated data from Generalized gd model, with seed 6
#'
#' A dataset containing the allele counts and genotypes from the Generalized gd model.
#' See (Salehi, Sohrab, Adi Steif, Roth Andrew, Samuel Aparicio, Alexandre Bouchard, and Sohrab P. Shah. “Joint Statistical Inference of Clonal Populations ￼from Single Cell and Bulk Tumour Sequencing Data.” (submitted).)
#'
#' @format A list of 8 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'   \item phi: For internal use of the package.
#'   \item trueMutMatrix: For internal use of the package.
#'   \item psis.single.cell: For internal use of the package.
#'   \item gdDirName: For internal use fo the package.
#'  }
#'
"gd.10.48.6.f0.gl0.u.tips.ADO.00.doublet.00.m.50.nus.0.da"


#' Simulated data from Generalized gd model, with seed 7
#'
#' A dataset containing the allele counts and genotypes from the Generalized gd model.
#' See (Salehi, Sohrab, Adi Steif, Roth Andrew, Samuel Aparicio, Alexandre Bouchard, and Sohrab P. Shah. “Joint Statistical Inference of Clonal Populations ￼from Single Cell and Bulk Tumour Sequencing Data.” (submitted).)
#'
#' @format A list of 8 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'   \item phi: For internal use of the package.
#'   \item trueMutMatrix: For internal use of the package.
#'   \item psis.single.cell: For internal use of the package.
#'   \item gdDirName: For internal use fo the package.
#'  }
#'
"gd.10.48.7.f0.gl0.u.tips.ADO.00.doublet.00.m.50.nus.0.da"


#' Simulated data from Generalized gd model, with seed 8
#'
#' A dataset containing the allele counts and genotypes from the Generalized gd model.
#' See (Salehi, Sohrab, Adi Steif, Roth Andrew, Samuel Aparicio, Alexandre Bouchard, and Sohrab P. Shah. “Joint Statistical Inference of Clonal Populations ￼from Single Cell and Bulk Tumour Sequencing Data.” (submitted).)
#'
#' @format A list of 8 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'   \item phi: For internal use of the package.
#'   \item trueMutMatrix: For internal use of the package.
#'   \item psis.single.cell: For internal use of the package.
#'   \item gdDirName: For internal use fo the package.
#'  }
#'
"gd.10.48.8.f0.gl0.u.tips.ADO.00.doublet.00.m.50.nus.0.da"


#' Simulated data from Generalized gd model, with seed 9
#'
#' A dataset containing the allele counts and genotypes from the Generalized gd model.
#' See (Salehi, Sohrab, Adi Steif, Roth Andrew, Samuel Aparicio, Alexandre Bouchard, and Sohrab P. Shah. “Joint Statistical Inference of Clonal Populations ￼from Single Cell and Bulk Tumour Sequencing Data.” (submitted).)
#'
#' @format A list of 8 variables:
#' \itemize{
#'   \item mutCounts: An integer matrix, with columns corresponding to the genomic loci. First and second rows contain the reference and variant alelle counts respectively.
#'   \item psi: The copy number at this genomic locus, encoded as a list, each element of which is a list of 3, \code{gV}, \code{gN}, and \code{gR},
#'    corresponding to variant Copy Number (CN), reference genotype CN, and normal CN.
#'    \code{gV} is a string in an alphabet of {A,B}, where the number of {A}-s denotes the major and the number of {B}-s denotes the minor CN respectively.
#'   \item filteredMutMatrix: A binary genotype-locus matrix, with element [i,j] is one if genotype i is mutated at locus j and is zero otherwise.
#'   \item mutPrevalence: For internal use of the package.
#'   \item missingIndx: For internal use of the package.
#'   \item phi: For internal use of the package.
#'   \item trueMutMatrix: For internal use of the package.
#'   \item psis.single.cell: For internal use of the package.
#'   \item gdDirName: For internal use fo the package.
#'  }
#'
"gd.10.48.9.f0.gl0.u.tips.ADO.00.doublet.00.m.50.nus.0.da"




