
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <title>defuse/NEWS | Shah Lab</title>
    
    <meta name="author" content="Sohrab P. Shah">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>

    <!-- Bootstrap styles -->
    <link href="/assets/themes/lab/bootstrap/css/bootstrap.min.css" rel="stylesheet"/>
    <!-- Optional theme -->
    <link href="/assets/themes/lab/bootstrap/css/bootstrap-theme.min.css" rel="stylesheet"/>
    <!-- Sticky Footer -->
    <link href="/assets/themes/lab/bootstrap/css/bs-sticky-footer.css" rel="stylesheet"/>
    
    <!-- Custom styles -->
    <link href="/assets/themes/lab/css/style.css?body=1" rel="stylesheet" type="text/css" media="all"/>

    <!-- icon -->
    <link rel="icon" type="image/png" href="/assets/themes/lab/images/logo/favicon.png"/>

    <!-- Custom fonts and icons via Font Awesome, http://fortawesome.github.io/Font-Awesome/ -->
    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css"/>

    <!-- Fonts via Google -->
    <link href='http://fonts.googleapis.com/css?family=Lato:300italic,700italic,300,700' rel='stylesheet' type='text/css'/>

    <!-- Math via MathJax -->
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
      <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->

    <!-- Fav and touch icons -->
    <!-- Update these with your own images
      <link rel="shortcut icon" href="images/favicon.ico">
      <link rel="apple-touch-icon" href="images/apple-touch-icon.png">
      <link rel="apple-touch-icon" sizes="72x72" href="images/apple-touch-icon-72x72.png">
      <link rel="apple-touch-icon" sizes="114x114" href="images/apple-touch-icon-114x114.png">
    -->

    <!-- atom & rss feed
    <link href="/atom.xml" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="/rss.xml" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">
    -->

  </head>

  <body>

  <!-- Static top navbar -->

  <nav class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class='container'>
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="navbar-header">
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#jb-navbar-collapse">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a href="/">
        <img class="pull-left logo" src="/assets/themes/lab/images/logo/lab-logo.svg" type="image/svg+xml">
        </a>
      </div>

      <!-- Collect the nav links, forms, and other content for toggling -->
      <div class="collapse navbar-collapse" id="jb-navbar-collapse">
        <ul class="nav navbar-nav navbar-left hidden-sm">
          <!-- Don't put site title on home page -- there's already a big title there. -->
          
          
          
          <a class="navbar-brand" href="/">Shah Lab</a>
          
        </ul>
        <ul class="nav navbar-nav navbar-right pull-right">
          
          
          


  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/research">Research</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/blog/">Blog</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/papers/">Papers</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/projects/">Software</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/team/">Team</a></li>
      	
      
    
  



          <!--<li class="dropdown nav navbar-nav">-->
            <!--<a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">-->
              <!--More <span class="caret"></span>-->
            <!--</a>-->

            <!--<ul class="dropdown-menu nav navbar-nav" role="menu">-->
              <!---->
              <!---->
              <!--


  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  


-->
            <!--</ul>-->
          <!--</li>-->
        </ul>
      </div><!-- /.navbar-collapse -->
    </nav>
   </div>

   <!-- Content -->
   <div class="container">
      
<div class="page-header">
  <div class="title media-heading">defuse/NEWS</div>
  
</div>

<div class="row post-full">
  <div class="col-xs-12">
    <div class="date">
      <span></span>
    </div>
    <div class="content">
      <p><a href="https://github.com/amcpherson/defuse">Fork me on GitHub!</a></p>

<h1 id="news">News</h1>

<h2 id="version-080-is-now-available">Version 0.8.0 is now available</h2>

<p>Improvements:</p>

<ul>
  <li>Install via conda</li>
  <li>Removed paths from config</li>
  <li>Specify output files on the command line (courtesy Sergey Mitrofanov)</li>
  <li>Date and time logged for each step (courtesy Sergey Mitrofanov)</li>
</ul>

<p>Bugfixes:</p>
<ul>
  <li>Samtools command line fixes (courtesy Dusan Randjelovic)</li>
</ul>

<h2 id="version-062-is-now-available">Version 0.6.2 is now available</h2>

<p>Bugfixes:</p>

<ul>
  <li>Updated to work with newer version of gmap</li>
</ul>

<h2 id="version-061-is-now-available">Version 0.6.1 is now available</h2>

<p>Improvements:</p>

<ul>
  <li>deFuse now takes 2 fastq files as input, replacing the previous archaic input method</li>
  <li>Reimplemented <code class="highlighter-rouge">interrupted_indexN</code> and <code class="highlighter-rouge">splicing_indexN</code></li>
  <li>Optionally compute <code class="highlighter-rouge">interrupted_indexN</code> and <code class="highlighter-rouge">splicing_indexN</code> using <code class="highlighter-rouge">calculate_extra_annotations</code></li>
</ul>

<h2 id="version-060-is-now-available">Version 0.6.0 is now available</h2>

<p>Improvements:</p>

<ul>
  <li>Automatic downloading of reference genome and other required files</li>
  <li>Much improved speed due to use of GMAP for the majority of post-process alignments</li>
  <li>Switched back to adaboost as described in the original paper, the svm was occasionally classifying some highly expressed fusions as false</li>
</ul>

<p>Bugfixes:</p>

<ul>
  <li>fixed issue with <code class="highlighter-rouge">merge_read_stats.pl</code></li>
</ul>

<p>Other Changes:</p>

<ul>
  <li>no longer filtering garbage reads since it was slow, and generally not useful</li>
  <li><code class="highlighter-rouge">--mm</code> bowtie option for memory mapped bowtie index was removed from default options, add to <code class="highlighter-rouge">bowtie_options</code> if in the config if needed</li>
</ul>

<p>Comments:</p>

<ul>
  <li>The results for some longer read datasets are greatly improved by trimming a variable number of low quality bases from the ends of each read</li>
</ul>

<h2 id="version-050-is-now-available">Version 0.5.0 is now available</h2>

<p>Fixes the following bugs:</p>

<ul>
  <li>Removed overzealous warning: <code class="highlighter-rouge">Fastq error, unable to interpret readid</code></li>
  <li>fixed findcandidatereads crashes</li>
  <li>Increased parallelism for faster runtime, especially during split read alignment and spanning read clustering</li>
  <li>removed obsolete break_predict column from output</li>
</ul>

<p>No dataset rebuild required.</p>

<h2 id="version-043-is-now-available">Version 0.4.3 is now available</h2>

<p>Added a simulated test dataset and deFuse output based on default parameters <a href="http://sourceforge.net/projects/defuse/files/simulation/">here</a></p>

<p>Fixes the following bugs:</p>

<ul>
  <li>Inconsequential <code class="highlighter-rouge">Error: expected GTAG and GTC to be same size</code> message muted</li>
  <li>Removed dependence on human ensembl identifiers to allow for use of mouse genome/gene models</li>
  <li>Warning message when deFuse fails to import any sequence data</li>
  <li>Different read lengths for the same pair no longer fails</li>
  <li>Workaround for occasional samtools merge crash</li>
  <li>Wrong gene annotation bugfix</li>
  <li>Warning for small fragment lengths with discord_read_trim recommendation</li>
</ul>

<p>In addition, the script <code class="highlighter-rouge">get_reads.pl</code> has been added. Pass it the config filename, output directory, and <code class="highlighter-rouge">cluster_id</code> of a given fusion to view the supporting spanning and split reads.</p>

<p>No dataset rebuild required.</p>

<h2 id="version-042-is-now-available">Version 0.4.2 is now available</h2>

<p>Fixes the est island filter that was previously broken. Adds 50 A’s to each cDNA sequence to allow for alignment of poly A tail reads.</p>

<p>Requires a dataset rebuild using the <code class="highlighter-rouge">create_reference_dataset.pl</code> script.</p>

<h2 id="version-041-is-now-available">Version 0.4.1 is now available</h2>

<p>Version 0.4.1 fixes the <code class="highlighter-rouge">too many reference sequences</code> bug of when using version 0.4.0 with the latest ensembl.</p>

<p>To start 0.4.1 where 0.4.0 left off, delete <code class="highlighter-rouge">*cluster*</code> from the output directory and run 0.4.1 on that output directory. There is no need to rebuild the reference dataset.</p>

<h2 id="version-040-is-now-available">Version 0.4.0 is now available</h2>

<p>Version 0.4.0 provides several advantages over version 0.3 defuse:</p>

<ul>
  <li>Fusions between a gene and an intergenic region are now detected</li>
  <li>local realignment of predicted sequenced is used to reduce false positives</li>
  <li>blat alignment for annotation is split over multiple jobs for speed and lower memory use</li>
  <li>as with version 0.3.7, version 0.4.0 should avoid a combinatorial explosion in the number of clusters</li>
</ul>

<p>From version 0.4.0 onwards, a prepackaged dataset will be replaced by step by step instructions on how to build your own dataset. The 0.3 dataset will not be compatible with version 0.4.0 and onwards.</p>

<h2 id="version-037-is-now-available">Version 0.3.7 is now available</h2>

<p>Version 0.3.7 fixes speed and memory issues that would occur with some datasets. If deFuse was taking a long time and generating a very large clusters.txt file, this update should help. In order to run the new version on a dataset that failed with version 0.3.6 or lower, first remove all files matching <em>clusters</em> in the output directory created by the previous version, then restart version 0.3.7 using the old output directory. The new version should pick up where the old version left off.</p>

<p>Known issues:</p>

<ul>
  <li>blat may fail on low memory machines, to be fixed soon by splitting up the reference for blat</li>
</ul>

<h2 id="version-036-is-now-available">Version 0.3.6 is now available</h2>

<p>Version 0.3.6 introduces new annotations that are leveraged using the adaboost classifier for slightly higher accuracy. The filtered output is now based on the probability produced by the adaboost classifier. deFuse 0.3.6 updates can be used to quickly reannotate a deFuse 0.3.5 analysis, however the repeats.txt file from the newly posted dataset is required. To reannotate:</p>

<p><code class="highlighter-rouge">rm output_directory/annotations.txt</code> <code class="highlighter-rouge">./reannotate.pl -c config.txt -o output_directory</code></p>

<p>Known issues:</p>

<ul>
  <li>we are currently working on speed issues that may be a problem for some larger datasets</li>
  <li>blat may fail on low memory machines, to be fixed soon by splitting up the reference for blat</li>
</ul>

<h2 id="version-035-is-now-available">Version 0.3.5 is now available</h2>

<p>Version 0.3.5 is a minor update that fixes a number of small bugs and fixes <code class="highlighter-rouge">reannotate.pl</code> so that it can actually be used to classify the results of any defuse 0.3.X run. This functionality was not working for 0.3.4. Simply run the following to annotate any 0.3.X run and obtain the new annotations and adaboost probability score. It is not necessary to run <code class="highlighter-rouge">reannotate.pl</code> if you have already run version 0.3.5 of <code class="highlighter-rouge">defuse.pl</code>.</p>

<p><code class="highlighter-rouge">rm output_directory/annotations.txt</code> <code class="highlighter-rouge">./reannotate.pl -c config.txt -o output_directory -p max_threads</code></p>

<p>Known issues:</p>

<ul>
  <li>blat may fail on low memory machines, to be fixed soon by splitting up the reference for blat</li>
  <li>the method is somewhat sensitive to the max_insert_size parameter, set to 3 standard deviations above expected fragment length, but no higher</li>
</ul>

<h2 id="version-034-is-now-available">Version 0.3.4 is now available</h2>

<p>Version 0.3.4 uses an adaboost classifier trained on 60 true positives and 61 false positives to produce a single probability score for each fusion. The R ada package is required. You can take full advantage of the adaboost classifier for results produced using other 0.3.X versions of defuse by simply running the <code class="highlighter-rouge">reannotate.pl</code> script <strong>**Edit: this does not work, please update to version 0.3.5</strong>**. Doing a full rerun using 0.3.4 should not be necessary. Once again, the dataset has not changed.</p>

<p>Changes:</p>

<ul>
  <li>adaboost classifier to produce a single probability score for each prediction</li>
  <li>additional features calculated for each fusion, including boundary sequence di-nucleotide entropy and fusion boundary homology</li>
  <li>bugfix for long read lengths and short fragment lengths</li>
</ul>

<p>Known issues:</p>

<ul>
  <li>blat may fail on low memory machines, to be fixed soon by splitting up the reference for blat</li>
  <li>the method is somewhat sensitive to the max_insert_size parameter, set to 3 standard deviations above expected fragment length, but no higher</li>
</ul>

<h2 id="version-033-is-now-available">Version 0.3.3 is now available</h2>

<p>Version 0.3.3 reworks the split alignments so they are much quicker and do not hit the disk as much as for previous versions. If your deFuse runs are taking a long time, its a good idea to upgrade. The dataset has not changed.</p>

<p>Changes:</p>

<ul>
  <li>reworked split alignments to be quicker and less disk intensive</li>
  <li>discord_read_trim in config.txt allows better use of libraries with long read lengths and shorter fragment lengths</li>
  <li>quality type for bowtie is configurable by setting bowtie_quals in config.txt</li>
</ul>

<p>Known issues:</p>

<ul>
  <li>blat may fail on low memory machines, to be fixed soon by splitting up the reference for blat</li>
  <li>the method is somewhat sensitive to the max_insert_size parameter, set to 3 standard deviations above expected fragment length, but no higher</li>
</ul>

<h2 id="version-032-is-now-available">Version 0.3.2 is now available</h2>

<p>If you were having problems previously with deFuse creating very large files, consuming large amounts of memory and taking long amounts of time, it could be because version 0.3.1 and 0.3.0 were not properly filtering IG rearrangements. The problems are exacerbated if your RNA-Seq data was produced from a tumour with high amounts of immune infiltration. This issue is now fixed in version 0.3.2. Note that results from 0.3.1 and 0.3.0 will not be wrong, but will include a superset of what you’re interested in (assuming you are not interested in predicting IG rearrangements).</p>

<p>Version 0.3.2 also provides a number of other annotations that may be interesting for prioritizing fusions for validation or further experiments.</p>

<p>Changes:</p>

<ul>
  <li>fixed exclusion of IG rearrangements, thereby fixing overly high resource usage by deFuse</li>
  <li>denovo breakpoint assembly is now optional via a setting in config.txt</li>
  <li>annotation of the expression of each gene</li>
  <li>annotation of fusions spliced at exon boundarys corrected</li>
  <li>annotation of an interruption index added</li>
  <li>annotation of an fusion splice index added</li>
  <li>annotation of genomic position and strand of fusion splice added</li>
</ul>

<p>Known issues:</p>

<ul>
  <li>blat may fail on low memory machines, to be fixed soon by splitting up the reference for blat</li>
  <li>libraries with long read lengths and short fragment lengths (ie 75bp reads and 150 bp fragments) will not work, to be fixed by chopping large reads</li>
</ul>

<h2 id="version-031-is-now-available">Version 0.3.1 is now available</h2>

<p>Changes:</p>

<ul>
  <li>Two unnecessary and potentially large files were being produced, this is no longer the case</li>
</ul>

<h2 id="version-030-is-now-available">Version 0.3.0 is now available</h2>

<p>This version represents a major change to the way split reads are calculated and has produced more validated predicitons than 0.2.0.  Not backward compatible with 0.2.0.</p>

<p>Changes:</p>

<ul>
  <li>New split read calculation</li>
  <li>Parallelized split read calculation</li>
  <li>Output now includes column headers</li>
</ul>

<h2 id="version-020-is-now-available">Version 0.2.0 is now available</h2>

<p>This is the first official release of deFuse.</p>

    </div>

  <!-- 
    <ul class="tag_box inline">
      <li><i class="glyphicon glyphicon-open"></i></li>
      
      


  
    
  


    </ul>
    

  
    <ul class="tag_box inline">
      <li><i class="glyphicon glyphicon-tags"></i></li>
      
      


  
    
  



    </ul>
   -->  
  
    <hr>
    
  </div>
</div>


   </div>


  <!-- Static bottom navbar -->
  <nav class="navbar navbar-default navbar-fixed-bottom">
    <div class="container">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#jb-bottom-collapse"> <!-- NOTE! data-target was changed to jb-bottom-collapse -->
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
      </div>
      <!-- Collect the nav links, forms, and other content for toggling -->
      <div class="collapse navbar-collapse" id="jb-bottom-collapse">
        <a href=http://molonc.bccrc.ca>
          <img class="logo" alt="BCCA Molecular Oncology" src=/assets/images/bcca_logo.png>
        </a>
        <a href=http://www.ubc.ca/>
          <img class="logo" alt="UBC" src=/assets/images/4_2016_UBCLogo_BlueRGB72.png>
        </a>
        <ul class="nav navbar-nav navbar-right navbar-small">
          <li><a href='/about.html'>About</a></li>
          <li><a href='/join.html'>Join</a></li>
          <li><a href='/contact.html'>Contact</a></li>
        </ul>
      </div><!-- /.navbar-collapse -->
    </div>
  </nav>

    




  <script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-103919622-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>




    <!-- Latest compiled and minified JavaScript, requires jQuery 1.x (2.x not supported in IE8) -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
    <script src="/assets/themes/lab/bootstrap/js/bootstrap.min.js"></script>
    <script type='text/javascript' src='https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js'></script>
  </body>
</html>

