
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <title>apolloh | Shah Lab</title>
    
    <meta name="author" content="Sohrab P. Shah">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>

    <!-- Bootstrap styles -->
    <link href="/assets/themes/lab/bootstrap/css/bootstrap.min.css" rel="stylesheet"/>
    <!-- Optional theme -->
    <link href="/assets/themes/lab/bootstrap/css/bootstrap-theme.min.css" rel="stylesheet"/>
    <!-- Sticky Footer -->
    <link href="/assets/themes/lab/bootstrap/css/bs-sticky-footer.css" rel="stylesheet"/>
    
    <!-- Custom styles -->
    <link href="/assets/themes/lab/css/style.css?body=1" rel="stylesheet" type="text/css" media="all"/>

    <!-- icon -->
    <link rel="icon" type="image/png" href="/assets/themes/lab/images/logo/favicon.png"/>

    <!-- Custom fonts and icons via Font Awesome, http://fortawesome.github.io/Font-Awesome/ -->
    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css"/>

    <!-- Fonts via Google -->
    <link href='http://fonts.googleapis.com/css?family=Lato:300italic,700italic,300,700' rel='stylesheet' type='text/css'/>

    <!-- Math via MathJax -->
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
      <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->

    <!-- Fav and touch icons -->
    <!-- Update these with your own images
      <link rel="shortcut icon" href="images/favicon.ico">
      <link rel="apple-touch-icon" href="images/apple-touch-icon.png">
      <link rel="apple-touch-icon" sizes="72x72" href="images/apple-touch-icon-72x72.png">
      <link rel="apple-touch-icon" sizes="114x114" href="images/apple-touch-icon-114x114.png">
    -->

    <!-- atom & rss feed
    <link href="/atom.xml" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="/rss.xml" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">
    -->

  </head>

  <body>

  <!-- Static top navbar -->

  <nav class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class='container'>
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="navbar-header">
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#jb-navbar-collapse">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a href="/">
        <img class="pull-left logo" src="/assets/themes/lab/images/logo/lab-logo.svg" type="image/svg+xml">
        </a>
      </div>

      <!-- Collect the nav links, forms, and other content for toggling -->
      <div class="collapse navbar-collapse" id="jb-navbar-collapse">
        <ul class="nav navbar-nav navbar-left hidden-sm">
          <!-- Don't put site title on home page -- there's already a big title there. -->
          
          
          
          <a class="navbar-brand" href="/">Shah Lab</a>
          
        </ul>
        <ul class="nav navbar-nav navbar-right pull-right">
          
          
          


  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/research">Research</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/blog/">Blog</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/papers/">Papers</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/projects/">Software</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/team/">Team</a></li>
      	
      
    
  



          <!--<li class="dropdown nav navbar-nav">-->
            <!--<a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">-->
              <!--More <span class="caret"></span>-->
            <!--</a>-->

            <!--<ul class="dropdown-menu nav navbar-nav" role="menu">-->
              <!---->
              <!---->
              <!--


  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  


-->
            <!--</ul>-->
          <!--</li>-->
        </ul>
      </div><!-- /.navbar-collapse -->
    </nav>
   </div>

   <!-- Content -->
   <div class="container">
      
<div class="page-header">
  <div class="title media-heading">apolloh</div>
  
</div>

<div class="row post-full">
  <div class="col-xs-12">
    <div class="date">
      <span></span>
    </div>
    <div class="content">
      <p><a href="https://github.com/shahcompbio/apolloh">Fork me on GitHub!</a></p>

<h2 id="running-the-compiled-software">Running the compiled software</h2>
<p>There are 2 ways to run the compiled software: 1) executable or 2) shell script. These options are offered by the Matlab as a result of using the MCR compiler. If you have MCR already installed and added to your path (specifically the LD_LIBRARY_PATH environment variable) then you can use the executable; otherwise, use the shell script as it allows you to manually specify the MCR install path.</p>

<div class="language-shell highlighter-rouge"><pre class="highlight"><code><span class="c">### In Linux ###</span>
<span class="nb">cd</span> &lt;<span class="nv">$install_dir</span>&gt;/APOLLOH_0.1.1/bin/

<span class="c">### Option 1) Run APOLLOH: Executable ###</span>
<span class="nb">export </span><span class="nv">LD_LIBRARY_PATH</span><span class="o">=</span>&lt;<span class="nv">$install_dir</span>&gt;/MATLAB_Component_Runtime/v77/runtime/glnxa64/:&lt;<span class="nv">$install_dir</span>&gt;/MATLAB_Component_Runtime/v77/sys/os/glnxa64/:&lt;<span class="nv">$install_dir</span>&gt;/MATLAB_Component_Runtime/v77/bin/glnxa64/
./apolloh &lt;<span class="nv">$input_allelic_ratio_datafile</span>&gt; &lt;<span class="nv">$input_copy_number_datafile</span>&gt; ../parameters/K18/apolloh_K18_params_Illumina_stromalRatio_Hyper10k.mat &lt;<span class="nv">$output_converged_parameter_file</span>&gt; &lt;<span class="nv">$output_results_file</span>&gt;


<span class="c">### Option 2) Run APOLLOH: Shell Script ###</span>
./run_apolloh.sh &lt;<span class="nv">$install_dir</span>&gt;/MATLAB_Component_Runtime/v77/ &lt;<span class="nv">$input_allelic_ratio_datafile</span>&gt; &lt;<span class="nv">$input_copy_number_datafile</span>&gt; ../parameters/K18/apolloh_K18_params_Illumina_stromalRatio_Hyper10k.mat &lt;<span class="nv">$output_converged_parameter_file</span>&gt; &lt;<span class="nv">$output_results_file</span>&gt;


<span class="c">### Sort data by chromosome and position ###</span>
sort -k1,1n -k2,2n &lt;<span class="nv">$output_results_file</span>&gt; tmp.txt
mv tmp.txt &lt;<span class="nv">$output_results_file</span>&gt;

<span class="c">### Generate segment file ###</span>
../scripts/createSingleSegFileFromAPOLLOH.pl -i &lt;<span class="nv">$output_results_file</span>&gt; -o &lt;<span class="nv">$output_segment_results_file</span>&gt; -calls 1;
</code></pre>
</div>

<h2 id="example-running-the-compiled-software-on-test-data">Example: Running the compiled software on test data</h2>
<p>We have provided some data and a script to test that APOLLOH is properly installed and able to run. These files can also serve as a guide to preparing and formatting your input data files.</p>

<div class="language-shell highlighter-rouge"><pre class="highlight"><code><span class="gp">&gt; </span><span class="nb">cd test</span>/
<span class="gp">&gt; </span>head -5 cnfile.txt
ID chrom start end num.mark seg.mean state.num state.name
<span class="nb">test </span>1 1 340000 340 0.128103 3 NEUT
<span class="nb">test </span>1 340001 354000 14 0.672409 5 AMP
<span class="nb">test </span>1 354001 1509000 1155 0.135829 3 NEUT
<span class="nb">test </span>1 1509001 1542000 33 0.833131 6 HLAMP
<span class="nb">test </span>1 1542001 5288000 3746 0.137101 3 NEUT
<span class="gp">&gt; </span>head -5 infile.txt
1 1 N 27 N 3
1 698 N 16 N 3
1 4803 N 2 N 9
1 7652 N 37 N 6
1 7997 N 1 N 8
<span class="gp">&gt; </span>./run_APOLLOH_test.sh
</code></pre>
</div>

<h2 id="running-the-software-in-matlab">Running the software in Matlab</h2>
<p>If you have Matlab installed and wish to run within the Matlab environment, then you can start up Matlab and add the source files before executing the main function.</p>

<div class="highlighter-rouge"><pre class="highlight"><code>&gt; ### In Linux ###
&gt; cd &lt;$install_dir&gt;/APOLLOH_0.1.0/bin/
&gt; &lt;$matlab_dir&gt;/bin/matlab
</code></pre>
</div>

<pre><code class="language-Matlab">&gt;&gt; % In Matlab
&gt;&gt; addpath(genpath("&lt;$install_dir&gt;/APOLLOH_0.1.0/hmm"))
&gt;&gt; addpath(genpath("&lt;$install_dir&gt;/APOLLOH_0.1.0/util"))
&gt;&gt; % assuming your have all the arguments to the main function assigned...
&gt;&gt; apolloh(infile,cnfile,paramset,outparam,outfile)

</code></pre>

<h2 id="inputs-and-outputs">Inputs and Outputs</h2>
<p>When using the compiled executable or running within the Matlab environment, the same input/output files and parameters are required:</p>
<div class="language-matlab highlighter-rouge"><pre class="highlight"><code><span class="n">Function</span> <span class="n">apolloh</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span><span class="n">cnfile</span><span class="p">,</span><span class="n">paramset</span><span class="p">,</span><span class="n">outparam</span><span class="p">,</span><span class="n">outfile</span><span class="p">)</span>

<span class="n">INPUTS</span><span class="p">:</span>  
<span class="n">infile</span>         <span class="n">Tab</span><span class="o">-</span><span class="n">delimited</span> <span class="nb">input</span> <span class="n">file</span> <span class="n">containing</span> <span class="n">allelic</span> <span class="n">counts</span>
                 <span class="n">from</span> <span class="n">the</span> <span class="n">tumour</span> <span class="n">at</span> <span class="n">positions</span> <span class="n">determined</span> <span class="n">as</span> <span class="n">heterozygous</span>
                 <span class="n">from</span> <span class="n">the</span> <span class="n">normal</span> <span class="n">genome</span><span class="o">.</span> 
                 <span class="n">No</span> <span class="n">header</span> <span class="nb">line</span> <span class="nb">is</span> <span class="n">assumed</span><span class="o">.</span>
                <span class="mi">6</span> <span class="n">columns</span><span class="p">:</span>
                    <span class="mi">1</span><span class="p">)</span> <span class="n">chr</span> <span class="p">(</span><span class="n">integer</span><span class="p">;</span> <span class="s1">'X'</span> <span class="nb">and</span> <span class="s1">'Y'</span> <span class="n">strings</span> <span class="n">can</span> <span class="n">be</span> <span class="n">used</span><span class="p">)</span>
                    <span class="mi">2</span><span class="p">)</span> <span class="n">position</span>
                    <span class="mi">3</span><span class="p">)</span> <span class="n">reference</span> <span class="n">base</span> <span class="p">(</span><span class="n">can</span> <span class="n">be</span> <span class="n">arbitrary</span><span class="p">;</span> <span class="nb">not</span> <span class="n">used</span><span class="p">)</span>
                    <span class="mi">4</span><span class="p">)</span> <span class="n">referenc</span> <span class="n">count</span>
                    <span class="mi">5</span><span class="p">)</span> <span class="n">non</span><span class="o">-</span><span class="n">reference</span> <span class="n">base</span> <span class="p">(</span><span class="n">can</span> <span class="n">be</span> <span class="n">arbitrary</span><span class="p">;</span> <span class="nb">not</span> <span class="n">used</span><span class="p">)</span>
                    <span class="mi">6</span><span class="p">)</span> <span class="n">non</span><span class="o">-</span><span class="n">reference</span> <span class="n">count</span> 

<span class="n">cnfile</span>         <span class="n">Tab</span><span class="o">-</span><span class="n">delimited</span> <span class="nb">input</span> <span class="n">copy</span> <span class="n">number</span> <span class="n">segment</span> <span class="n">prior</span> <span class="n">file</span><span class="o">.</span>
                  <span class="n">The</span> <span class="n">accepted</span> <span class="nb">format</span> <span class="nb">is</span> <span class="n">the</span> <span class="n">output</span> <span class="n">from</span> <span class="n">HMMcopy</span><span class="p">,</span>
                  <span class="n">a</span> <span class="n">read</span><span class="o">-</span><span class="n">depth</span> <span class="k">for</span> <span class="n">analyzing</span> <span class="n">copy</span> <span class="n">number</span> <span class="n">in</span> <span class="n">tumour</span><span class="o">-</span>
                  <span class="n">normal</span> <span class="n">sequenced</span> <span class="n">genomes</span><span class="o">.</span>  
                  <span class="n">However</span><span class="p">,</span> <span class="n">copy</span> <span class="n">number</span> <span class="n">segments</span> <span class="n">from</span> <span class="nb">any</span> <span class="n">source</span> <span class="n">can</span> <span class="n">be</span> <span class="n">used</span><span class="o">.</span>                
                 <span class="mi">8</span><span class="o">-</span><span class="n">columns</span><span class="p">:</span>
                    <span class="mi">1</span><span class="p">)</span> <span class="n">id</span> <span class="p">(</span><span class="n">can</span> <span class="n">be</span> <span class="n">arbitrary</span><span class="p">,</span> <span class="nb">not</span> <span class="n">used</span><span class="p">)</span>
                    <span class="mi">2</span><span class="p">)</span> <span class="n">chr</span>
                    <span class="mi">3</span><span class="p">)</span> <span class="n">start</span>  
                    <span class="mi">4</span><span class="p">)</span> <span class="n">stop</span>
                    <span class="mi">5</span><span class="p">)</span> <span class="n">Number</span> <span class="n">of</span> <span class="mi">1</span><span class="n">kb</span> <span class="n">intervals</span> <span class="p">(</span><span class="n">can</span> <span class="n">be</span> <span class="n">arbitrary</span><span class="p">;</span> <span class="nb">not</span> <span class="n">used</span><span class="p">)</span>
                    <span class="mi">6</span><span class="p">)</span> <span class="nb">median</span> <span class="nb">log2</span> <span class="n">ratio</span> <span class="p">(</span><span class="n">normal</span> <span class="nb">and</span> <span class="n">tumour</span><span class="p">)</span> <span class="k">for</span> <span class="n">segment</span>
                    <span class="mi">7</span><span class="p">)</span> <span class="n">HMM</span> <span class="n">state</span><span class="p">:</span> <span class="mi">1</span><span class="o">=</span><span class="n">HOMD</span> <span class="p">(</span><span class="mi">0</span> <span class="n">copies</span><span class="p">),</span> <span class="mi">2</span><span class="o">=</span><span class="n">HEMD</span> <span class="p">(</span><span class="mi">1</span> <span class="n">copy</span><span class="p">),</span>
                        <span class="mi">3</span><span class="o">=</span><span class="n">NEUT</span> <span class="p">(</span><span class="mi">2</span> <span class="n">copies</span><span class="p">),</span> <span class="mi">4</span><span class="o">=</span><span class="n">GAIN</span> <span class="p">(</span><span class="mi">3</span> <span class="n">copies</span><span class="p">),</span>
                        <span class="mi">5</span><span class="o">=</span><span class="n">AMP</span> <span class="p">(</span><span class="mi">4</span> <span class="n">copies</span><span class="p">),</span> <span class="mi">6</span><span class="o">=</span><span class="n">HLAMP</span> <span class="p">(</span><span class="mi">5</span><span class="o">+</span> <span class="n">copies</span><span class="p">)</span>
                        <span class="n">Note</span> <span class="n">that</span> <span class="k">for</span> <span class="n">AMP</span> <span class="nb">and</span> <span class="n">HLAMP</span><span class="p">,</span> <span class="n">relative</span> <span class="n">numbers</span> <span class="n">of</span> <span class="n">copies</span>
                        <span class="n">can</span> <span class="n">be</span> <span class="n">used</span> <span class="p">(</span><span class="nb">i</span><span class="o">.</span><span class="n">e</span><span class="o">.</span> <span class="n">GAIN</span> <span class="nb">is</span> <span class="mi">3</span><span class="o">-</span><span class="mi">4</span> <span class="n">copies</span><span class="p">,</span> <span class="n">AMP</span> <span class="nb">is</span> <span class="mi">5</span><span class="o">-</span><span class="mi">6</span> <span class="n">copies</span><span class="p">,</span>
                        <span class="n">HLAMP</span> <span class="nb">is</span> <span class="mi">7</span><span class="o">+</span> <span class="n">copies</span><span class="p">)</span>
                    <span class="mi">8</span> <span class="p">)</span> <span class="n">CN</span> <span class="n">state</span> <span class="p">(</span><span class="n">can</span> <span class="n">be</span> <span class="n">arbitrary</span><span class="p">;</span> <span class="nb">not</span> <span class="n">used</span><span class="p">)</span>
                  <span class="n">If</span> <span class="n">cnfile</span><span class="o">=</span><span class="s1">'0'</span> <span class="nb">is</span> <span class="n">used</span><span class="p">,</span> <span class="n">then</span> <span class="n">copy</span> <span class="n">number</span> <span class="n">of</span> <span class="mi">2</span> <span class="p">(</span><span class="n">diploid</span><span class="p">)</span> <span class="nb">is</span> <span class="n">used</span>
                    <span class="k">for</span> <span class="nb">all</span> <span class="n">positions</span><span class="o">.</span>  

<span class="n">paramset</span>       <span class="n">Parameter</span> <span class="n">intialization</span> <span class="n">file</span> <span class="nb">is</span> <span class="n">a</span> <span class="n">matlab</span> <span class="n">binary</span> <span class="p">(</span><span class="o">.</span><span class="n">mat</span><span class="p">)</span> <span class="n">file</span><span class="o">.</span>
                      <span class="n">This</span> <span class="n">file</span> <span class="n">contains</span> <span class="n">model</span> <span class="nb">and</span> <span class="n">setting</span> <span class="n">paramters</span> <span class="n">necessary</span>
                      <span class="n">to</span> <span class="nb">run</span> <span class="n">the</span> <span class="n">program</span><span class="o">.</span>
                      <span class="n">See</span> <span class="n">examples</span> <span class="n">in</span> <span class="err">"</span><span class="o">&lt;</span><span class="err">$</span><span class="n">install_dir</span><span class="o">&gt;</span><span class="p">/</span><span class="n">APOLLOH_0</span><span class="o">.</span><span class="mf">1.0</span><span class="p">/</span><span class="n">parameters</span><span class="p">/</span><span class="err">"</span><span class="o">.</span>

<span class="n">outfile</span>        <span class="n">Tab</span><span class="o">-</span><span class="n">delimited</span> <span class="n">output</span> <span class="n">file</span> <span class="k">for</span> <span class="n">position</span><span class="o">-</span><span class="n">level</span> <span class="n">results</span><span class="o">.</span> 
                  <span class="mi">9</span><span class="o">-</span><span class="n">columns</span><span class="p">:</span>
                     <span class="mi">1</span><span class="p">)</span> <span class="n">chr</span> <span class="p">(</span><span class="s1">'X'</span> <span class="nb">and</span> <span class="s1">'Y'</span> <span class="n">will</span> <span class="n">be</span> <span class="n">output</span> <span class="n">as</span> <span class="mi">23</span> <span class="nb">and</span> <span class="mi">24</span><span class="p">)</span>
                     <span class="mi">2</span><span class="p">)</span> <span class="n">position</span>
                     <span class="mi">3</span><span class="p">)</span> <span class="n">reference</span> <span class="n">count</span>
                     <span class="mi">4</span><span class="p">)</span> <span class="n">non</span><span class="o">-</span><span class="n">reference</span> <span class="n">count</span>
                     <span class="mi">5</span><span class="p">)</span> <span class="n">total</span> <span class="n">depth</span>
                     <span class="mi">6</span><span class="p">)</span> <span class="n">allelic</span> <span class="n">ratio</span>
                     <span class="mi">7</span><span class="p">)</span> <span class="n">copy</span> <span class="n">number</span> <span class="p">(</span><span class="n">from</span> <span class="nb">input</span><span class="p">)</span>
                     <span class="mi">8</span> <span class="p">)</span> <span class="n">APOLLOH</span> <span class="n">genotype</span> <span class="n">state</span>
                     <span class="mi">9</span><span class="p">)</span> <span class="n">Zygosity</span> <span class="n">state</span><span class="o">.</span>
                  <span class="n">N</span> <span class="n">additional</span> <span class="n">columns</span><span class="p">:</span>
                     <span class="n">posterior</span> <span class="n">marginal</span> <span class="n">probabilities</span> <span class="p">(</span><span class="n">responsibilities</span><span class="p">)</span> <span class="k">for</span>
                     <span class="n">each</span> <span class="n">APOLLOH</span> <span class="n">genotype</span> <span class="n">state</span><span class="o">.</span> 
                 <span class="n">Zygosity</span> <span class="n">states</span> <span class="n">are</span><span class="p">:</span>
                    <span class="n">DLOH</span><span class="o">=</span><span class="n">deletion</span><span class="o">-</span><span class="n">LOH</span> <span class="p">(</span><span class="n">state</span> <span class="mi">1</span><span class="p">)</span>
                    <span class="n">NLOH</span><span class="o">=</span><span class="n">copy</span><span class="o">-</span><span class="n">neutral</span><span class="o">-</span><span class="n">LOH</span> <span class="p">(</span><span class="n">states</span> <span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">)</span>
                    <span class="n">ALOH</span><span class="o">=</span><span class="n">amplified</span><span class="o">-</span><span class="n">LOH</span> <span class="p">(</span><span class="n">states</span> <span class="mi">5</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">13</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">19</span><span class="p">)</span>
                    <span class="n">HET</span><span class="o">=</span><span class="n">heterozygous</span> <span class="p">(</span><span class="n">states</span> <span class="mi">3</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">)</span>
                    <span class="n">ASCNA</span><span class="o">=</span><span class="n">allele</span><span class="o">-</span><span class="n">specific</span><span class="o">-</span><span class="n">amplification</span> <span class="p">(</span><span class="n">states</span> <span class="mi">10</span><span class="p">,</span><span class="mi">12</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">18</span><span class="p">)</span>
                    <span class="n">BCNA</span><span class="o">=</span><span class="n">balanced</span><span class="o">-</span><span class="n">amplification</span> <span class="p">(</span><span class="n">states</span> <span class="mi">11</span><span class="p">,</span><span class="mi">16</span><span class="p">,</span><span class="mi">17</span><span class="p">)</span>

                  <span class="n">Segment</span> <span class="n">boundaries</span> <span class="n">are</span> <span class="n">determined</span> <span class="n">as</span> <span class="n">consecutive</span>
                   <span class="n">marginal</span> <span class="n">states</span> <span class="n">of</span> <span class="n">DLOH</span><span class="p">,</span> <span class="n">NLOH</span><span class="p">,</span> <span class="n">ALOH</span><span class="p">,</span> <span class="n">HET</span><span class="p">,</span> <span class="n">BCNA</span><span class="p">,</span>
                   <span class="n">ASCNA</span><span class="p">;</span> <span class="n">this</span> <span class="n">implementation</span> <span class="n">does</span> <span class="nb">not</span> <span class="n">output</span> <span class="n">this</span>
                   <span class="n">information</span><span class="o">.</span> <span class="n">An</span> <span class="n">external</span> <span class="n">Perl</span> <span class="n">script</span> <span class="n">handles</span> <span class="n">this</span><span class="p">:</span> <span class="err">"</span>
                   <span class="o">&lt;</span><span class="err">$</span><span class="n">install_dir</span><span class="o">&gt;</span><span class="p">/</span><span class="n">APOLLOH_0</span><span class="o">.</span><span class="mf">1.0</span><span class="p">/</span><span class="n">scripts</span><span class="p">/</span><span class="n">createSingleSegFileFromAPOLLOH</span><span class="o">.</span><span class="n">pl</span><span class="err">"</span>

<span class="n">outparam</span>       <span class="n">Tab</span><span class="o">-</span><span class="n">delimited</span> <span class="n">output</span> <span class="n">file</span> <span class="n">storing</span> <span class="n">converged</span> <span class="n">parameters</span>
                       <span class="n">after</span> <span class="n">model</span> <span class="n">training</span> <span class="n">using</span> <span class="n">Expectation</span> <span class="n">Maximization</span> <span class="p">(</span><span class="n">EM</span><span class="p">)</span>
                       <span class="n">algorithm</span><span class="o">.</span>
                     <span class="mi">1</span><span class="p">)</span> <span class="n">Number</span> <span class="n">of</span> <span class="n">iterations</span> 
                     <span class="mi">2</span><span class="p">)</span> <span class="n">Global</span> <span class="n">normal</span> <span class="n">contamination</span> <span class="n">parameter</span> 
                     <span class="mi">3</span><span class="p">)</span> <span class="n">Binomial</span> <span class="n">parameters</span> <span class="k">for</span> <span class="n">each</span> <span class="n">HMM</span> <span class="nb">class</span><span class="p">/</span><span class="n">state</span><span class="o">.</span>
</code></pre>
</div>

    </div>

  <!-- 
    <ul class="tag_box inline">
      <li><i class="glyphicon glyphicon-open"></i></li>
      
      


  
    
  


    </ul>
    

  
    <ul class="tag_box inline">
      <li><i class="glyphicon glyphicon-tags"></i></li>
      
      


  
    
  



    </ul>
   -->  
  
    <hr>
    
  </div>
</div>


   </div>


  <!-- Static bottom navbar -->
  <nav class="navbar navbar-default navbar-fixed-bottom">
    <div class="container">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#jb-bottom-collapse"> <!-- NOTE! data-target was changed to jb-bottom-collapse -->
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
      </div>
      <!-- Collect the nav links, forms, and other content for toggling -->
      <div class="collapse navbar-collapse" id="jb-bottom-collapse">
        <a href=http://molonc.bccrc.ca>
          <img class="logo" alt="BCCA Molecular Oncology" src=/assets/images/bcca_logo.png>
        </a>
        <a href=http://www.ubc.ca/>
          <img class="logo" alt="UBC" src=/assets/images/4_2016_UBCLogo_BlueRGB72.png>
        </a>
        <ul class="nav navbar-nav navbar-right navbar-small">
          <li><a href='/about.html'>About</a></li>
          <li><a href='/join.html'>Join</a></li>
          <li><a href='/contact.html'>Contact</a></li>
        </ul>
      </div><!-- /.navbar-collapse -->
    </div>
  </nav>

    




  <script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-103919622-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>




    <!-- Latest compiled and minified JavaScript, requires jQuery 1.x (2.x not supported in IE8) -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
    <script src="/assets/themes/lab/bootstrap/js/bootstrap.min.js"></script>
    <script type='text/javascript' src='https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js'></script>
  </body>
</html>

